git/pre-commit

```diff
@@ -0,0 +1,80 @@
+#!/bin/sh
+### Colors
+BGBLUE='\033[44m'
+RED='\033[91m'
+GREEN="\e[32m"
+NC='\033[0m' # No Color
+
+### Styles
+NORMAL=$(tput sgr0)
+BOLD=$(tput bold)
+
+### Outputs
+abort_commit() {
+   printf "${RED}Commit aborted 🛑 ${NORMAL} \n"
+}
+
+success_commit() {
+  printf "${GREEN} All good ✅ ${NORMAL} \n"
+  exit 0
+}
+
+check_errors() {
+    let "ERROR = $1"
+    if [ "${ERROR}" -ne "0" ]
+    then
+        abort_commit
+        exit "${ERROR}"
+    fi
+}
+
+check_required_commands() {
+    if ! command -v "$1" >/dev/null 2>&1;
+    then
+        printf "${RED}${BOLD}$1${NORMAL}${RED} is required but it's not installed. Install it please, and then try committing again ${NORMAL} \n"
+        abort_commit
+        exit 1
+    fi
+}
+
+#### START PRE-COMMIT HOOK ###
+echo "${BGBLUE}Running pre-commit checks on your code...${NORMAL}"
+for COMMAND in "golangci-lint" "hadolint" "trivy"; do
+    check_required_commands "${COMMAND}"
+done
+
+#### golangci-lint
+echo "${BGBLUE}>> golangci-lint run${NC}"
+golangci-lint run
+check_errors $?
+
+#### Format
+FILES=$(go list ./...  | grep -v mock)
+echo "${BGBLUE}>> go fmt${NC}"
+go fmt ${FILES}
+
+#### Modules
+echo "${BGBLUE}>> go mod tidy${NC}"
+go mod tidy
+
+#### Check all files for suspicious constructs
+echo "${BGBLUE}>> go vet${NC}"
+go vet ${FILES}
+check_errors $?
+
+#### Launching tests
+echo "${BGBLUE}>> go test -count=1 -race ./...${NC}"
+go test -count=1 -race ./...
```

Comment:
does this mean that we have to run all the tests before committing? Maybe it is too hard if there are many tests?

Reply:
Yes, that runs all tests. My idea is to run just unit tests in the pre-commit hook (it's all we have for now), so it shouldn't take more than a few seconds. This way you'd know if your changes have broken some tests and you wouldn't have to wait for the CI to tell you